icu4c/source/i18n/messageformat2_function_registry.cpp

```diff
@@ -421,25 +423,185 @@ static FormattedPlaceholder notANumber(const FormattedPlaceholder& input) {
     return FormattedPlaceholder(input, FormattedValue(UnicodeString("NaN")));
 }
 
-static FormattedPlaceholder stringAsNumber(const number::LocalizedNumberFormatter& nf, const FormattedPlaceholder& input, UErrorCode& errorCode) {
+// Returns true if `c` is in the interval [`first`, `last`]
+static bool inRange(UChar32 c, UChar32 first, UChar32 last) {
+    U_ASSERT(first < last);
+    return c >= first && c <= last;
+}
+
+static bool isDigit(UChar32 c) { return inRange(c, 0x0030, 0x0039); }
+
+static int32_t parseDigit(UChar32 c) {
+    switch(c) {
+    case u'0': {
+        return 0;
+    }
+    case u'1': {
+        return 1;
+    }
+    case u'2': {
+        return 2;
+    }
+    case u'3': {
+        return 3;
+    }
+    case u'4': {
+        return 4;
+    }
+    case u'5': {
+        return 5;
+    }
+    case u'6': {
+        return 6;
+    }
+    case u'7': {
+        return 7;
+    }
+    case u'8': {
+        return 8;
+    }
+    case u'9': {
+        return 9;
+    }
+    default: {
+        // Should be unreachable
+        return 0;
+    }
+    }
+}
+
+static int32_t parseDigits(const UnicodeString& s, int32_t& i) {
+    int32_t firstDigit = i;
+    // Assumes that i < s.length() and s[i] is a digit
+    while (isDigit(s[i])) {
+        i++;
+    }
+    int32_t numDigits = i - firstDigit;
+    int32_t placeValue = 1;
+    int32_t result = 0;
+    for (int32_t j = firstDigit + numDigits - 1; j >= firstDigit; j--) {
+        result += placeValue * parseDigit(s[j]);
+        placeValue *= 10;
+    }
```

Comment:
simpler way to parse a stringified decimal number from front to back rather than back to front
```suggestion
    int32_t result = 0;
    while (isDigit(s[i])) {
        result *= 10;
        result += parseDigit(s[i]);
        i++;
    }
```